scala-ql/src/main/scala/scalaql/SideEffect.scala

```diff
@@ -1,38 +1,61 @@
 package scalaql
 
-final case class SideEffect[R, S, A](
-  initialState: S,
-  acquire:      () => R,
-  release:      (R, S) => Unit,
-  use:          (R, S, A) => S) { self =>
+final class SideEffect[R, S, A] private (
```

Comment:
Thoughts on using cats-effect Resource? https://typelevel.org/cats-effect/docs/std/resource

Reply:
@Swoorup I appreciate your review! 
I'm trying to avoid bringing IO monads and tagless-final stuff to this library because it will slightly increase the complexity.
You can take a look at what could happen in case of adding cats IO 🤣 
The example is from my studentship project

https://github.com/itkpi/trembita/blob/master/kernel/src/main/scala/trembita/DataPipelineT.scala